README.md
# hoe-debugging

* http://github.com/jbarnette/hoe-debugging


## Description

A Hoe plugin to help you debug your C extensions. This plugin provides `test:gdb` and `test:valgrind` tasks (plus a few variants). As of v1.5 it also can generate a valgrind suppression from a previous test suite's log file.

See the `Hoe::Debugging` module for a few configuration options.

This plugin expects you to have `gdb` and `valgrind` available in your `PATH`.


## Summary

In your Rakefile:

``` ruby
Hoe.plugin :debugging
```

Then you'll get the following rake tasks:

```
rake test:valgrind               # debugging # Run the test suite under Valgrind
rake test:valgrind:mem           # debugging # Run the test suite under Valgrind with memory-fill
rake test:valgrind:mem0          # debugging # Run the test suite under Valgrind with memory-zero
rake test:valgrind:suppression   # debugging # Generate a valgrind suppression file for your test suite
rake valgrind:suppression[file]  # debugging # Generate a valgrind suppression file from a previous run's log file
```


## Examples

Run your test suite under gdb:

``` sh
rake test:gdb
```

Run your test suite with valgrind's memcheck:

``` sh
rake test:valgrind
```

If you have repeatable valgrind warnings that you've decided it's OK to suppress:

``` sh
rake test:valgrind:suppression
rake test:valgrind # when this runs, the previous run's errors will be suppressed
```

If you have a log file containing hard-to-reproduce valgrind warnings (e.g., from CI) that you've decided it's OK to suppress:

``` sh
rake valgrind:suppression[path/to/file]
rake test:valgrind # when this runs, the errors from that log file will be suppressed
```


## Suppression files

Suppression files can be added to the `suppressions` subdirectory of your project.

As of v2.0.0, `hoe-debugging` will use all suppression files that match any part of your Ruby's version. It uses the Hoe project name and the ruby version name, with an optional trailing `/_.*/`.

For example, if:

* my Hoe project was named `myproject`
* and my Ruby version was `2.5.1.57` (the `57` is `RUBY_PATCHLEVEL`)

then the following would be found and used:

* `suppressions/myproject_ruby-2.5.1.57.supp`
* `suppressions/myproject_ruby-2.5.1.57_namespace_deref.supp`
* `suppressions/myproject_ruby-2.5.1.supp`
* `suppressions/myproject_ruby-2.5.supp`
* `suppressions/myproject_ruby-2.supp`
* `suppressions/myproject_ruby-2_exit_frees.supp`

and the following would **not** be used:

* `suppressions/otherproject_ruby-2.5.1.57.supp` because the project name is wrong
* `suppressions/myproject_ruby-2.5.1.58.supp` because the patchlevel is wrong
* `suppressions/myproject_ruby-2.5.2.supp` because the patch is wrong
* `suppressions/myproject_ruby-2.4.supp` because the minor is wrong
* `suppressions/myproject_ruby-1.supp` because the major is wrong


## Installation

```
$ gem install hoe-debugging
```

You should also use your distro's package manager to install `gdb` and `valgrind`.


## License

MIT. See `LICENSE` file in this repository.


CHANGELOG.md
## Changelog for `hoe-debugging`

### 2.0.0 / 2018-11-04

Features:

* Allow multiple suppression files to be active at once. Pull in all
  suppression files that match a part of the ruby version. Along with
  this change, the helper method `matching_suppression_file` is
  renamed to `matching_suppression_files`.


### 1.5.0 / 2018-11-02

Features:

* Suppression metadata is always logged when warnings are generated
* Rake task `valgrind:suppression` can generate a suppression file from a previous run's log file. Great for when CI finds things that are hard to reproduce!


### 1.4.2 / 2017-06-19

Features:

* Allow matching suppression file on major Ruby version.


### 1.4.1 / 2017-06-19

Features:

* Support frozen string constants.


### 1.4.0 / 2017-06-19

Features:

* Add explanatory messages to suppression file generation.


Bugfixes:

* Fixed generation of suppression files.


### 1.3.0 / 2017-01-20

* The rake task now fails if valgrind detects any errors during the run.


### 1.2.1 / 2015-12-16

Features:

* Set larger stack size (via `ulimit -s`) to properly support Ruby 2.1 and later


### 1.2.0 / 2015-01-22

Features:

* Ruby 2.2 support.


Bugfixes:

* Replace references to `Config` with `RbConfig`, for Ruby 2.2 support.


### 1.1.1 / 2012-11-26

Bugfixes:

* Correctly name suppression files for REE.


### 1.1.0 / 2012-09-26

* Valgrind suppressions can be generated with `rake test:valgrind:suppression`
* If present, valgrind suppressions will be used when valgrinding.

### 1.0.5 / 2012-06-19

* RSpec support now works with TESTOPTS.

### 1.0.4 / 2012-06-19

* Adding support for RSpec.

### 1.0.3 / 2012-02-06

* Fixing compatibility with latter-day versions of rake. #1

### 1.0.2 / 2011-11-13

* Fixing reference to ::RUBY with rbconfig-based construction of the ruby interpreter path.

### 1.0.1 / 2009-06-26

* Fix Hoe dependency version.
* Activate the hoe-git plugin.
* Fix docs, give Mike some well-deserved credit!

### 1.0.0 / 2009-06-22

* Birthday!


spec/spec_helper.rb
require File.join File.dirname(__FILE__), "../lib/hoe/debugging"
require "fileutils"
require "rake"


test/test_hoe_debugging.rb
require "test/unit"
require "hoe_debugging"

class TestHoeDebugging < Test::Unit::TestCase
  def test_sanity
    flunk "write tests or I will kneecap you"
  end
end


spec/hoe_debugging_exit_code_spec.rb
require_relative "spec_helper.rb"

describe "exit code" do
  def sh command
    Rake.sh command, verbose: false
  end

  def do_in_test_dir
    Bundler.with_clean_env do
      Dir.chdir test_dir do
        saved_gemfile = ENV['BUNDLE_GEMFILE']
        ENV['BUNDLE_GEMFILE'] = "./Gemfile"
        begin
          yield
        ensure
          ENV['BUNDLE_GEMFILE'] = saved_gemfile
        end
      end
    end
  end

  let(:test_dir) { File.join(File.dirname(__FILE__), "files/test_project") }

  before do
    do_in_test_dir do
      begin
        sh "bundle install > #{logfile} 2>&1"
        sh "bundle exec rake compile >> #{logfile} 2>&1"
      rescue RuntimeError => e
        puts "SETUP FAILED:\n#{File.read(logfile)}"
        raise
      end
    end
  end

  context "from a good run" do
    let(:logfile) { "good-run.log" }
    it "is zero" do
      do_in_test_dir do
        system "bundle exec rake test:valgrind TESTOPTS='--name /notexist/' >> #{logfile} 2>&1"
        exitcode = $?
        expect(exitcode.success?).to(be_truthy, File.read(logfile))
      end
    end
  end

  context "from a bad run" do
    let(:logfile) { "bad-run.log" }
    it "is nonzero" do
      do_in_test_dir do
        system "bundle exec rake test:valgrind >> #{logfile} 2>&1"
        exitcode = $?
        expect(exitcode.success?).to(be_falsey, File.read(logfile))
      end
    end
  end
end


lib/hoe/debugging.rb
require 'rbconfig'
require 'tempfile'

class Hoe #:nodoc:

  # Whee, stuff to help when your codes are b0rked. Tasks provided:
  #
  # * <tt>test:gdb</tt>
  # * <tt>test:valgrind</tt>
  # * <tt>test:valgrind:mem</tt>
  # * <tt>test:valgrind:mem0</tt>

  module Debugging
    VERSION = "2.0.0" #:nodoc:

    ##
    # The exit code of valgrind when it detects an error.
    ERROR_EXITCODE = 42 # the answer to life, the universe, and segfaulting.

    ##
    # Optional: Used to add flags to GDB. [default: <tt>[]</tt>]

    attr_accessor :gdb_options

    ##
    # Optional: Used to add flags to valgrind. [default:
    # <tt>%w(--num-callers=50 --error-limit=no --partial-loads-ok=yes
    # --undef-value-errors=no)</tt>]

    attr_accessor :valgrind_options

    def initialize_debugging #:nodoc:
      self.gdb_options = []

      self.valgrind_options = ["--num-callers=50",
                               "--error-limit=no",
                               "--partial-loads-ok=yes",
                               "--undef-value-errors=no",
                               "--error-exitcode=#{ERROR_EXITCODE}",
                               "--gen-suppressions=all",
                              ]
    end

    def hoe_debugging_ruby
      # http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-talk/151376
      @ruby ||= File.join(RbConfig::CONFIG["bindir"], (RbConfig::CONFIG["RUBY_INSTALL_NAME"] + RbConfig::CONFIG["EXEEXT"]))
    end

    def hoe_debugging_make_test_cmd
      cmd = []
      if File.directory? "spec"
        cmd << "-S rspec"
        cmd << (ENV['FILTER'] || ENV['TESTOPTS'])
      else
        cmd << make_test_cmd
      end
      cmd.join(' ')
    end

    def hoe_debugging_command
      "#{hoe_debugging_ruby} #{hoe_debugging_make_test_cmd}"
    end

    def hoe_debugging_run_valgrind command, cmdline_options=[]
      sh "#{hoe_debugging_valgrind_helper.valgrind} #{cmdline_options.join(' ')} #{command}"
    end

    def hoe_debugging_check_for_suppression_file options
      suppression_files = hoe_debugging_valgrind_helper.matching_suppression_files
      suppression_files.each do |suppression_file|
        puts "NOTICE: using valgrind suppressions in #{suppression_file.inspect}"
        options << "--suppressions=#{suppression_file}"
      end
    end

    def hoe_debugging_valgrind_helper
      @valgrind_helper ||= ValgrindHelper.new name
    end

    def define_debugging_tasks #:nodoc:
      desc "Run the test suite under GDB."
      task "test:gdb" do
        sh "gdb #{gdb_options.join ' '} --args #{hoe_debugging_command}"
      end

      desc "Run the test suite under Valgrind."
      task "test:valgrind" do
        vopts = valgrind_options
        hoe_debugging_check_for_suppression_file vopts
        hoe_debugging_run_valgrind hoe_debugging_command, vopts
      end

      desc "Run the test suite under Valgrind with memory-fill."
      task "test:valgrind:mem" do
        vopts = valgrind_options + ["--freelist-vol=100000000", "--malloc-fill=6D", "--free-fill=66"]
        hoe_debugging_check_for_suppression_file vopts
        hoe_debugging_run_valgrind hoe_debugging_command, vopts
      end

      desc "Run the test suite under Valgrind with memory-zero."
      task "test:valgrind:mem0" do
        vopts = valgrind_options + ["--freelist-vol=100000000", "--malloc-fill=00", "--free-fill=00"]
        hoe_debugging_check_for_suppression_file vopts
        hoe_debugging_run_valgrind hoe_debugging_command, vopts
      end

      desc "Generate a valgrind suppression file for your test suite."
      task "test:valgrind:suppression" do
        vopts = valgrind_options
        generated_suppression_file = false
        ::Tempfile.open "hoe_debugging_valgrind_suppression_log" do |logfile|
          begin
            hoe_debugging_run_valgrind "#{hoe_debugging_ruby} #{hoe_debugging_make_test_cmd} 2> #{logfile.path}", vopts
          rescue RuntimeError
            suppfile = hoe_debugging_valgrind_helper.save_suppressions_from logfile.path
            puts "NOTICE: saved suppressions to #{suppfile}"
            generated_suppression_file = true
          end
        end
        unless generated_suppression_file
          puts "WARNING: no valgrind warnings detected, no suppressions file generated"
        end
      end

      desc "Generate a valgrind suppression file from a previous run's log file"
      task "valgrind:suppression", [:file] do |task, args|
        file_path = args[:file]
        raise "Please specify the path to your log file" unless file
        suppfile = hoe_debugging_valgrind_helper.save_suppressions_from file_path
        puts "NOTICE: saved suppressions to #{suppfile}"
      end
    end

    class ValgrindHelper
      DEFAULT_DIRECTORY_NAME = "suppressions"

      attr_accessor :directory, :project_name

      def initialize project_name, options={}
        @project_name = project_name
        @directory    = options[:directory] || DEFAULT_DIRECTORY_NAME
      end

      def formatted_ruby_version
        engine = if defined?(RUBY_DESCRIPTION) && RUBY_DESCRIPTION =~ /Ruby Enterprise Edition/
                   "ree"
                 else
                   defined?(RUBY_ENGINE) ? RUBY_ENGINE : "ruby"
                 end
        %Q{#{engine}-#{RUBY_VERSION}.#{RUBY_PATCHLEVEL}}
      end

      def version_matches
        matches = [formatted_ruby_version]                          # e.g. "ruby-2.5.1.57"
        matches << formatted_ruby_version.split(".")[0,3].join(".") # e.g. "ruby-2.5.1"
        matches << formatted_ruby_version.split(".")[0,2].join(".") # e.g. "ruby-2.5"
        matches << formatted_ruby_version.split(".")[0,1].join(".") # e.g. "ruby-2"
        matches
      end

      def parse_suppressions_from logfile_name
        suppressions = []
        File.open logfile_name do |logfile|
          suppression = nil
          while ! logfile.eof? do
            line = logfile.readline
            if suppression.nil?
              if line =~ /\<insert_a_suppression_name_here\>/
                suppression = String.new "{\n"
                suppression << line
              end
            else
              suppression << line
              if line =~ /^\}$/
                suppressions << suppression
                suppression = nil
              end
            end
          end
        end
        suppressions.uniq.join
      end

      def save_suppressions_from logfile, options={}
        ensure_directory_exists
        suppressions = parse_suppressions_from logfile
        filename = File.join directory, "#{project_name}_#{formatted_ruby_version}.supp"
        File.open(filename, "w") { |out| out.write suppressions }
        filename
      end

      def matching_suppression_files
        matching_files = []
        version_matches.each do |version_string|
          matching_files += Dir[File.join(directory, "#{project_name}_#{version_string}.supp")]
          matching_files += Dir[File.join(directory, "#{project_name}_#{version_string}_*.supp")]
        end
        matching_files
      end

      def valgrind
        # note that valgrind will generally crap out on rubies >= 2.1
        # unless we increase the max stack size beyond the default
        "ulimit -s unlimited && valgrind"
      end

      private

      def ensure_directory_exists
        return if File.directory? directory
        FileUtils.mkdir_p directory
        File.open(File.join(directory, "README.txt"), "w") do |readme|
          readme.puts "This directory contains valgrind suppression files generated by the hoe-debugging gem."
        end
      end
    end
  end
end
